src/components/QuickOpenModal.js

```diff
@@ -0,0 +1,378 @@
+    }
+
+    const nowEnabled = !prevProps.enabled && this.props.enabled;
+    const queryChanged = prevProps.query !== this.props.query;
+    if (nowEnabled || queryChanged) {
+      this.updateResults(this.props.query);
+    }
+  }
+
+  closeModal = () => {
+    this.props.closeQuickOpen();
+    this.props.clearHighlightLineRange();
+  };
+
+  searchSources = (query: string) => {
+    if (query == "") {
+      const results = this.props.sources;
+      this.setState({ results });
+      return;
+    }
+    const { searchType } = this.props;
+
+    if (searchType === "gotoSource") {
+      const [baseQuery] = query.split(":");
+      const results = filter(this.props.sources, baseQuery, { key: "value" });
+      this.setState({ results });
+    } else {
+      const results = filter(this.props.sources, query, { key: "value" });
+      this.setState({ results });
+    }
+  };
+
+  searchSymbols = (query: string) => {
+    const { symbols: { functions, variables }, searchType } = this.props;
+
+    let results = functions;
+    if (searchType === "variables") {
+      results = variables;
+    }
+    if (query == "" || query == "@" || query == "#") {
+      this.setState({ results });
+      return;
+    }
+
+    results = filter(results, query.slice(1), {
+      key: "value"
+    });
+
+    this.setState({ results });
+  };
+
+  updateResults = (query: string) => {
+    const { searchType } = this.props;
+    switch (searchType) {
+      case "functions":
+      case "variables":
+        return this.searchSymbols(query);
+      case "sources":
+      case "gotoSource":
+      default:
+        return this.searchSources(query);
+    }
+  };
+
+  selectResultItem = (
+    e: SyntheticEvent<HTMLElement>,
+    item: ?FormattedSource | ?FormattedSymbolDeclaration
+  ) => {
+    const { selectSource, selectedSource, query, searchType } = this.props;
+
+    switch (searchType) {
+      case "functions":
+      case "variables":
+        if (selectedSource == null || item == null) {
+          return;
+        }
+        selectSource(selectedSource.get("id"), {
+          location: {
+            ...(item.location != null ? { line: item.location.start.line } : {})
+          }
+        });
+        break;
+      case "gotoSource":
+        if (item == null) {
+          return;
+        }
+
+        const [, line, column] = query.split(":");
+        const lineNumber = parseInt(line, 10);
+        const columnNumber = parseInt(column, 10);
+        if (!isNaN(lineNumber)) {
+          selectSource(item.id, {
+            location: {
+              line: lineNumber,
+              ...(!isNaN(columnNumber) ? { column: columnNumber } : null)
+            }
+          });
+          break;
+        }
+        break;
+      case "sources":
+      default:
+        if (item == null) {
+          return;
+        }
+
+        selectSource(item.id);
+        break;
+    }
+
+    this.closeModal();
+  };
```

Comment:
so this might be a style thing, but the `case` statements and length of the function feel a bit much:

perhaps some helper methods and early returns?

```js 
 if (item == null) {
   return
 }

if() {
  return this.selectSymbol()
}
if () {
  this.selectSource()
`

selectSource(item.id);

```